Arnarayanan xsoarwebserver

arnarayanan opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A simple web server that can be used to host forms and handle user clicks. Contains an integration that can be used to setup the server to accept responses from users and poll the result and example playbooks and automations.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [x] 6.2.0
- [x] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

arnarayanan commented:
https://github.com/demisto/dockerfiles/pull/8234

MosheEichler reviewed (changes requested):
Hi @issolahma.
You did an Excellent Job!! :)

Please remove the following files.

> - Packs/XsoarWebserver/Integrations/XSOARWebServer/command_examples
> - Packs/XsoarWebserver/Integrations/XSOARWebServer/test_data/baseintegration-dummy.json
> 

Please run the [format](https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/format/README.md) command for the following file.

> - Packs/XsoarWebserver/Integrations/XSOARWebServer/XSOARWebServer.yml

Please run the [generate_docs](https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/generate_docs/README.md) command.


Please notify me when you are done.
Thanks!!

arnarayanan commented:
Hi @MosheEichler  Have made the suggested changes.
Please check and let me know

MosheEichler commented:
@arnarayanan Yes from my side is this ok let wait for @idovandijk to take a look and we are ready to go

arnarayanan commented:
@MosheEichler Will you approve the merge on the demisto/dockerfiles as well ?

MosheEichler commented on Packs/XsoarWebserver/Integrations/XSOARWebServer/XSOARWebServer.py:
```diff
@@ -0,0 +1,117 @@
+        if command == 'long-running-execution':
+            run_long_running(listen_port)
+        elif command == 'test-module':
+            pass
```
```suggestion
            pass
```
Why do you no testing the connection?

MosheEichler commented on Packs/XsoarWebserver/Integrations/XSOARWebServer/README.md:
```diff
@@ -0,0 +1,113 @@
+This is a simple web-server that as of now, supports handling configurable user responses (like Yes/No/Maybe). What makes it different from Data collection tasks is that, the URL to perform a certain action is predictable and written to the incident context when an action is setup. This URL can be inserted to for eg: an HTML email.  User clicks are  are recorded in the integration context and can ...
+This integration was integrated and tested with version 6.6 of XSOAR, but is expected to work from XSOAR version 6.2 and with and on 6.1 with Generic polling (instead of using the automation).
```
```suggestion
This integration was integrated and tested with version 6.6 of XSOAR, but is expected to work from XSOAR version 6.2 and on 6.1 with Generic polling (instead of using the automation).
```

MosheEichler commented on Packs/XsoarWebserver/Integrations/XSOARWebServer/README.md:
```diff
@@ -0,0 +1,113 @@
+This is a simple web-server that as of now, supports handling configurable user responses (like Yes/No/Maybe). What makes it different from Data collection tasks is that, the URL to perform a certain action is predictable and written to the incident context when an action is setup. This URL can be inserted to for eg: an HTML email.  User clicks are  are recorded in the integration context and can ...
```
```suggestion
This is a simple web-server that as of now, supports handling configurable user responses (like Yes/No/Maybe). What makes it different from Data collection tasks is that, the URL to perform a certain action is predictable and written to the incident context when an action is setup. This URL can be inserted to for eg: an HTML email. User clicks are recorded in the integration context and can be polled by Scheduled Commands / Generic Polling.
```

MosheEichler reviewed (changes requested):
Hi [arnarayanan](https://github.com/arnarayanan),
You did an excelent job :)

Please see my comments.
Please run demisto-sdk format -i Packs/XsoarWebserver/Integrations/XSOARWebServer/XSOARWebServer.yml

Let me know when you are done, Thanks!!

MosheEichler commented:
> @MosheEichler Will you approve the merge on the demisto/dockerfiles as well ?

Yes I'll go over it and approve it

MosheEichler commented:
The docker pr https://github.com/demisto/dockerfiles/pull/8234

arnarayanan commented on Packs/XsoarWebserver/Integrations/XSOARWebServer/XSOARWebServer.py:
```diff
@@ -0,0 +1,117 @@
+from bottle import route, run
+
+
+@route('/hello')
+def hello():
+    return "Hello World!"
```
i put it there as a way to check if the server is responding (if you want to check from the outsideO and if everything is alright. Can remove it, if it is against the standards

MosheEichler commented:
@arnarayanan regarding the docker image pr some files are missing, please [see](https://github.com/demisto/dockerfiles/blob/master/README.md) how to contribute a docker image.
Please let me know if you need some help

arnarayanan commented:
@MosheEichler  could you please explain which files are absolutely necessary ? for the docker repo. As of now it is only one dependency and hence the pip install inside the dockerfile
I can see that the build has succeeded using the Dockerfile

MosheEichler commented:
@arnarayanan Regarding the docker pr I'll take care onit

arnarayanan commented:
i think i have all the files you need for Docker on the repo Do you know if @idovandijk has seen the request
@MosheEichler

idovandijk reviewed (commented):
@arnarayanan Thanks for your contribution.
Please look at the Email Communication pack, it seems to do what you're doing here.
Let me know if there's anything I'm missing.

bakatzir commented:
What an effort 🥇 
I'm happy to take the feedback on what was missing from the EmailCommunication pack that made you commit to this solution.
We can discuss it, and take decisions accordingly.

Bar.

arnarayanan commented:
Email communication does not work for all the customers (for eg: some do not want their recipients to hit reply and answer a certain question.) Sometimes the emails are combined into a conversation and the recipients have to remove the whole thread and add just the reply (which is cumbersome). 

The ideal solution for my customer (for whom i built this and they have accepted this integration) was to include a simple hyper link on an email with an attachment and the recipient should click on it and their playbook would proceed accordingly. An acceptable solution would have been if there was a way to add an attachment to ask user task 
![image](https://user-images.githubusercontent.com/76943923/181524222-5bee7a49-a85f-42f5-81e3-736ab281b76b.png)
and the recipient can reply yes/no. However for my customer even this was suboptimal, since they did not want users to type a reply.
Another way to deal with it is if the Datacollection task link was made available in the context, then we could try and insert it in an email with an attachment, but that would not be as clean as this integration.

On another note, collecting user responses is just one of the use-cases made possible with this; hence the comparison with Email Communication is inaccurate @idovandijk  
I can imagine a use case where we can setup something like a polling mechanism where multiple user responses can be collected and a decision can be made (to branch to a part of the playbook) based on the result of responses of multiple users.
@bakatzir

idovandijk commented:
@arnarayanan I appreciate your work and detailed response.
I can see that Data Collection indeed doesn't allow attachments when sending the email.
But regarding what you wrote later - we already have an implementation of a data collection task that is sent to multiple users with the playbook taking different paths. It's implemented in a playbook loop in the Crisis Management pack which happens to be a pack I released.

I do acknowledge the gap that your pack fills. I'm asking 2 things before we proceed:
1. Can you please attach a demo video or schedule us a short demo demonstrating the usage?
2. Can you please give the playbooks easier to read names that meet our conventions? Example: `XSOAR Web Server - Send Email With Attachment`.

arnarayanan commented:
Hey @idovandijk  Did not know about the Crisis management pack. I had a quick look at it.
We can schedule a demo whenever you are free,  If you let me know some potential time slots, I could send an invite to whoever is attending.
Also will change the playbook names if required.

MosheEichler commented:
> @arnarayanan Regarding the docker pr I'll take care on it

This is the new docker pr
https://github.com/demisto/dockerfiles/pull/8284

idovandijk commented:
I sent out a demo invite for Thursday. Feel free to suggest other times if needed.

idovandijk commented:
@arnarayanan would u like to release the update that you were planning in this PR or do we close one and u open a new one?

arnarayanan commented:
@idovandijk i have replaced the code on this PR and removed the older files. So what we have here now is an entirely new code base. This has all the new features. 
We can go over it when you are free; please do send an invite at a convenient time. @meital-aizen

arnarayanan commented:
@MosheEichler can you help me figure out what is wrong with the file
Packs/XsoarWebserver/IncidentFields/incidentfield-XsoarWebserverEmailUserSLA.json 
This is the validation check that is failing

idovandijk commented:
@arnarayanan please send an invite in an open time-slot for me and @MosheEichler and @meital-aizen (whoever wants/could join, will).

arnarayanan commented:
have sent one for tomorrow

idovandijk reviewed (changes requested):
Thank you @arnarayanan for investing more time into improving this. After our demo here our my comments:
Your content is working, now the idea is to make it easy to use and easy to customize where possible. I understand some of this is simply examples for users to see - but let's make it easy for them to run and understand so that they can create custom versions for themselves. Fixes:
- Listname - let's make it a playbook input that accepts HTML with a default value of the template you've created. Users will then be able to replace the HTML completely with a list value like ${lists.myhtml}.
- Try using the counter incident field in a context key instead of an incident field. It should be initialized in the parent playbook, passed as an input to the subplaybook, incremented in the subplaybook, and finally outputted to the parent playbook. The loop should run until the counter has reached the desired number (in the loop settings).
- "Actions" should be playbook inputs (happy,angry)
- The subplaybook cannot be used in parallel with the same playbook because of "incident.emailsendcounter". See if there's a way to overcome this issue, or otherwise explain the playbook description that that limitation exists.
- Change playbook names to reflect better their usage - use Title Case and spaces and make the name related to XXSOAR Web Server. Example: `XSOAR Web Server - Email Data Collection` or `Web Server - Data Collection With Attachments`, or something else if you like.
- Please add playbook images to README files of the playbooks.
- I recommend adding to the pack description the advantages of this over the normal Email Ask or the normal Email Communication. I can foresee lots of people asking about what the difference is. Even better - if you are willing to - you can maybe add a short video to the pack README which would demonstrate its usage. This is simply a recommendation and not a must for this PR.

LMK if you hit any issues, and please send me the screenshots showing the "On Error" bug, with your XSOAR version. Thanks!

arnarayanan commented:
Listname - let's make it a playbook input that accepts HTML with a default value of the template you've created. Users will then be able to replace the HTML completely with a list value like ${lists.myhtml}. -> done

Try using the counter incident field in a context key instead of an incident field. It should be initialized in the parent playbook, passed as an input to the subplaybook, incremented in the subplaybook, and finally outputted to the parent playbook. The loop should run until the counter has reached the desired number (in the loop settings). -> not possible to do it this way, i ll have to set context to global for the subplaybook; which in my opinion is bad practice.

"Actions" should be playbook inputs (happy,angry) -> done


The subplaybook cannot be used in parallel with the same playbook because of "incident.emailsendcounter". See if there's a way to overcome this issue, or otherwise explain the playbook description that that limitation exists. -> **will  have to introduce a json incident field that keeps track of each job in the incident and an automation that updates this incident field. Will need more development and testing. Will do it in the next iteration.**

Change playbook names to reflect better their usage - use Title Case and spaces and make the name related to XXSOAR Web Server. Example: XSOAR Web Server - Email Data Collection or Web Server - Data Collection With Attachments, or something else if you like. ->done

Please add playbook images to README files of the playbooks. -> done

I recommend adding to the pack description the advantages of this over the normal Email Ask or the normal Email Communication. I can foresee lots of people asking about what the difference is. Even better - if you are willing to - you can maybe add a short video to the pack README which would demonstrate its usage. This is simply a recommendation and not a must for this PR. -> Have explained the basic construct in the main README file.

MosheEichler commented on Packs/XsoarWebserver/Integrations/XSOARWebServer/XSOARWebServer.py:
```diff
@@ -0,0 +1,251 @@
+    try:
+        data = integration_context[uuid]
+        data["link_tracker"][linkuuid]["response_received"] = True
+        data["link_tracker"][linkuuid]["response"] = dict(request.forms.decode())
```
```suggestion
        data["link_tracker"][linkuuid]["response"] = dict(request.forms.decode())  # pylint: disable=E1101
```

MosheEichler commented on Packs/XsoarWebserver/IncidentFields/incidentfield-emailsendcounter.json:
```diff
@@ -0,0 +1,30 @@
+    "editForm": true,
+    "group": 0,
+    "hidden": false,
+    "id": "incident_xsoar-web-serveremailsendcounter",
```
```suggestion
    "id": "incident_xsoarwebserveremailsendcounter",
```

MosheEichler commented on Packs/XsoarWebserver/IncidentFields/incidentfield-emailsendcounter.json:
```diff
@@ -0,0 +1,30 @@
+    "id": "incident_xsoar-web-serveremailsendcounter",
+    "isReadOnly": false,
+    "locked": false,
+    "name": "xsoar-web-serveremailsendcounter",
```
```suggestion
    "name": "Xsoar-web-server email send counter",
```

MosheEichler commented on Packs/XsoarWebserver/IncidentFields/incidentfield-XsoarWebserverEmailUserSLA.json:
```diff
@@ -0,0 +1,31 @@
+    "id": "incident_xsoar-web-serveremailusersla",
+    "isReadOnly": true,
+    "locked": false,
+    "name": "xsoar-web-serverEmailUserSLA",
```
```suggestion
    "name": "Xsoar-web-server Email User SLA",
```

MosheEichler commented on Packs/XsoarWebserver/IncidentFields/incidentfield-XsoarWebserverEmailUserSLA.json:
```diff
@@ -0,0 +1,31 @@
+    "editForm": false,
+    "group": 0,
+    "hidden": false,
+    "id": "incident_xsoar-web-serveremailusersla",
```
```suggestion
    "id": "incident_xsoarwebserveremailusersla",
```

idovandijk reviewed (commented):
Excellent job @arnarayanan, you've worked very hard on this, I can see it.
@MosheEichler this is approved from my side. If you don't require additional changes feel free to approve the PR.

MosheEichler commented:
@idovandijk Please approve it. i can't erge the pr until you approve it
